plugins/oidc/access.go

```diff
@@ -10,118 +10,207 @@ import (
+
+// discoverEndpoints retrieves the userinfo and jwks endpoints from the given
+// discovery endpoint.
+// The function sends a GET request to the discovery endpoint, which should
+// respond with the openid connect configuration.
+// This configuration is a JSON object that the function decodes into a map,
+// extracting the userinfo_endpoint and jwks_uri fields as the userinfo and jwks
+// endpoints respectively.
+// In case of error, it will also join the error with suitable contextual error
+// message and return.
+// If the discovery endpoint is empty or leads to a failed request, the function
+// returns an error.
+// If the retrieval or decoding of the openid connect configuration fails, the
+// function returns an error.
+//
+// Parameters:
+//
+//	discoveryEndpoint (string): address of the discovery endpoint.
+//
+// Returns:
+//
+//	userinfoEndpoint (string): the discovered userinfo endpoint.
+//	jwksEndpoint (string): the discovered jwks endpoint.
+//	err (error): the error, if any occurred during the process.
+func discoverEndpoints(discoveryEndpoint string) (userinfoEndpoint string, jwksEndpoint string, err error) {
+	// cleanup the discovery endpoint address and remove possible whitespaces
+	discoveryEndpoint = strings.TrimSpace(discoveryEndpoint)
+	if discoveryEndpoint == "" {
+		return "", "", errEmptyEndpoint
+	}
+
+	// now request the openid connect configuration endpoint
+	res, err := http.Get(discoveryEndpoint)
+	if err != nil {
+		return "", "", errors.Join(errDiscoveryRequestFailure, err)
+	}
+
+	// now read the openid connect configuration
+	var openIdConnectConfiguration map[string]interface{}
+	err = json.NewDecoder(res.Body).Decode(&openIdConnectConfiguration)
+	if err != nil {
+		return "", "", errors.Join(err, errDiscoveryUnmarshalFailure)
+	}
+
+	return openIdConnectConfiguration["userinfo_endpoint"].(string), openIdConnectConfiguration["jwks_uri"].(string), nil
+}
+
+// extractBearerToken is a function that takes an authorization header value as
+// input and extracts the bearer token from it.
+// It uses a regular expression to match the header value against a predefined
+// pattern.
+// If a match is found and there are two capture groups in the match result, the
+// function returns the second capture group, which represents the bearer token.
+// If no match is found or there are not exactly two capture groups, an empty
+// string is returned.
+//
+// Example usage:
+//
+//	token := extractBearerToken("Bearer abcxyz123")
+//	// token = "abcxyz123"
+//
+// Parameters:
+//   - authorizationHeaderValue: The value of the authorization header.
+//
+// Returns:
+//
+//	The extracted bearer token, or an empty string if no valid bearer token is found.
+func extractBearerToken(authorizationHeaderValue string) string {
+	regex := regexp.MustCompile(headerRegex)
+	matches := regex.FindStringSubmatch(authorizationHeaderValue)
+	if len(matches) == 3 {
+		return matches[2]
+	}
+	return ""
```

Comment:
Why return an empty string instead of `nil`. I think a `nil` would be clearer. (Or some optional type.)

Reply:
Returning `nil` would require using `*string` as return type and requires a more thorough check of the return type in the code later on, since a `*string` cannot be used directly in a `if` statement to check the length and would require the following syntax making the function using the return value longer:
```go
if accessToken == nil {
  // raise error here
}

if strings.Trim(*accessToken) == "" {
  // raise another error here
}
```